### ISSUE TITLE
Alerts

### ISSUE THREAD

**Issue by benfrancis** [mozilla-iot/gateway#1295]
As a user I want to be notified when a door is opened or motion is detected in my home so that I can act on that information.

**Comment by benfrancis** [comment]
The immediate solution that comes to mind for this is to add a notification output to the rules engine:
![notification_output](https://user-images.githubusercontent.com/552417/44590238-94f68000-a7b2-11e8-9411-1bcecb5834fb.png)

Users could then set a rule which sends a push notification with whatever message they want (using the [Push API](https://developer.mozilla.org/en-US/docs/Web/API/Push_API) and [Notifications API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API)) in the web app.

There are a couple of downsides I can think of with this approach:
1. I'm not sure how reliable push notifications in web apps are across different desktop and mobile platforms, and cross-browser support is still not great
2. Arming and disarming alerts for your security system (motion sensors, door sensors, leak sensors etc.) would require manually turning rules on and off every time you leave and arrive at your house

Regarding 1, we could in future add email and SMS notifications as alternatives in the drop-down menu on the rule output, but that will require stepping up some back end infrastructure to make those functions work. (Maybe in future we might also want to consider building a native app in addition to the web app, which has native push notifications).

Regarding 2, you could probably work around this by using an and statement with a virtual on/off switch which switches all your alerts on and off. But that is a hack which most users are very unlikely to think of.

An alternative to this design is a dedicated security area of the UI with a global armed/un-armed state, a bit like the Smart Home Monitor that SmartThings has. That is probably much more intuitive.

What we could do is add the simple notification rule output as in the mockup above, and then gather feedback about how people use it in order to design a more dedicated UI later.

Thoughts?

**Comment by rzr** [comment]
Based on a few experiments on Social WebOfThings
I had in mind to make an ActivityPub adapter, let user to

1/ Enter user credentials of online account (ie: mastodon)
2/ define message template (like you mocked up)

Then things will be wired using the rule engine, without additional efforts.

More insights at:

https://github.com/rzr/webthing-iotjs/wiki/Social

**Comment by kgiori** [comment]
I would think about how a physical security system works. When you leave, you arm it, when you return, you type a code to disarm it. Now imagine that type of simple security system controller box as a web thing. We should add a description of it by growing the [schemas](https://iot.mozilla.org/schemas/) list. 

"SecurityMonitor" capability
* Properties: boolean (armed/disarmed) - required, sound (alarm sound via tone? freq? siren?) - optional; strobe (some form of illumination alert) - optional, announce (recorded or live voice feed?) - optional
* Actions: alarm (sets a sound, strobe, or voice feed to occur for a period of X seconds) - optional, listen (uses a mic to enable live stream of audio, or begin recording) - optional, camera (snapshots every x seconds or video stream enabled or recording) - optional
* Events: SecurityAlert (if any security devices monitored by "armed" state are triggered, set off this event) - optional

The minimum required boolean property of armed/disarmed could feed the input to a rule that triggers a **notification**. When you leave you say "[Turn on | Arm] the <name of SecurityMonitor>". There could be several options for disarming the system, to make it hard to disable for strangers, but easy for kids coming home from school before you.

Then if we add a Virtual Security Monitor to our Virtual Things add-on, you don't actually need a physical box -- you can "arm" and "disarm" your home using the virtual thing. You can create a rule with inputs to it from door, window, motion, sound, etc. sensors.

In summary, if security sensors fire, rules can trigger virtual monitor "Security Alert" events.
If the virtual monitor is not armed, those events could be logged but no action would be taken. If the virtual monitor is armed **and** a "Security Alert" event occurs, then you can trigger a notification in a rule, as well as trigger other alarms.

**Comment by benfrancis** [comment]
@hobinjk I think we should go ahead and implement a simple text notification as a rule output to fulfill this use case. (I say simple, getting push notifications to work may be tricky, and I'm not sure how cross-platform or reliable it's going to be).

Here is the notification icon: https://drive.google.com/file/d/1VWW4IOFih72dvdy20uqyUiMWaDqers6Y/view?usp=sharing

Do you need anything else to get started implementing this?

**Comment by benfrancis** [comment]
UI mockup (as above): https://docs.google.com/presentation/d/1rMeg6hep8JQflqwbiNIbhSoK9StCeVZ2npphmQHKda4/edit#slide=id.g40d79a403a_0_5

**Comment by hobinjk** [comment]
This is definitely doable, I'll get started tomorrow now that I think I won't be sick then. It might end up a bit tricky to get the notification event from the rules engine back to the browser and may require all the bells and whistles of the Push API.

- [ ] Add notification rule part block UI
- [ ] Hook this up to the Notification API
- [ ] Figure out the Push API

**Comment by hobinjk** [comment]
Currently it looks like the best way of getting notifications from the server to the client will be a websocket carrying event data. This could be rolled into the main Gateway as a proper WebThing API "event" or it could be left as a weird offshoot of the rules engine.

**Comment by benfrancis** [comment]
> Currently it looks like the best way of getting notifications from the server to the client will be a websocket carrying event data.

Surely that will only work if the web interface is currently loaded in a browser tab? That isn't really fit for purpose to inform me of an intruder in my house.

The idea of using push notifications with Service Workers is that once installed they should should still work even if the browser tab is closed (and on mobile, even when the browser app or installed PWA is closed).

Have you explored this option?

**Comment by hobinjk** [comment]
Oh right, sorry. I got distracted by the implications of making the notifications webthing api events. The best way is definitely to implement the full Push API.

**Comment by benfrancis** [comment]
Yeah to be clear they shouldn't be Web Thing API events. A notification is a special type of output for the rules engine (like we have the clock input), which sends a push notification. That way a rule can be set up to trigger a notification based on an property, action or event as an input (e.g. if motion is true on a motion sensor or open is true on a door sensor).

In future we can add other types of notification like email or SMS in a dropdown.

**Comment by hobinjk** [comment]
I agree that for the time being it should just be a magic notification generation block with no other side effects.

Moving forwards, I think it's interesting to represent features of the gateway as part of the web thing API. In the case of notifications, this would mean that visiting the gateway's `/events` endpoint or event log naturally gives a list of all notifications it has recently generated. If the clock input were represented as a virtual on/off switch (hidden from the UI by default) it would be easy to debug time-based rules.

The ideal future I have in mind is that the rules engine contains a minimal amount of magic in its implementation and instead does as much as possible within the web thing api.

**Comment by benfrancis** [comment]
> Moving forwards, I think it's interesting to represent features of the gateway as part of the web thing API. 

I agree, the gateway itself is a device. We could represent the physical state of LEDs as properties, reboot as an action and button presses as events for example. There's a question as to how far we take that though, there's no need to represent all internal state and all software features of the gateway via the Web Thing API. Probably something to follow up on in a separate issue/issues later.
